File: sorting/bucket_sort/bucketSort.java
// Sort a large set of floating point numbers which are in the range from 0.0 to 1.0

public class bucketSort {

    public static void doBucketSort(double a[]) {
        int size = a.length;
        int index[] = new int[10];
        double b[][] = new double[10][size];
        for (int i = 0; i < size; i++) {
            int number = (int) (a[i] * 10);
            int j;
            for (j = 0; j < index[number]; j++) {
                if (b[number][j] > a[i]) {
                    break;
                }
            }
            double pre = b[number][j];
            b[number][j] = a[i];
            for (int k = j + 1; k < index[number] + 1; k++) {
                double temp = b[number][k];
                b[number][k] = pre;
                pre = temp;

            }
            index[number]++;
        }
        int count = 0;
        for (int i = 0; i < 10; i++) {
            if(index[i]!=0){
                for (int j = 0; j < index[i]; j++) {
                    a[count] = b[i][j];
                    count++;
                }
            }

        }
    }


    public static void main(String[] args) {

        double a[] = {0.1, 0.23, 0.0, 0.50, 0.12, 0.99, 0.56, 0.2, 0.1, 0.11};
        System.out.println("Unsorted Array: ");
        for (int i = 0; i < a.length; i++) {
            System.out.print(a[i] + " ");
        }
        System.out.println();
        doBucketSort(a);
        System.out.println("Sorted Array: ");
        for (int i = 0; i < a.length; i++) {
            System.out.print(a[i] + " ");
        }
    }
}

File: sorting/bucket_sort/bucket_sort.cpp
// Sort a large set of floating point numbers which are in the range from 0.0 to 1.0

#include<iostream>

using namespace std;

void counting_sort(double a[], int n){
	double b[10][n];
	int index[10];

	for(int i=0; i<10; i++){
		index[i] = 0;
	}

	for (int i =0; i<n; i++){
		int number = a[i]*10;
		int j;
		for(j=0; j<index[number]; j++){
			if(b[number][j] >= a[i]){
				break;
			}
		}
		double pre = a[i];
		for(int k = j; j<index[number]; j++){
			double temp = b[number][j];
			b[number][j] = pre;
			pre = temp;
		}
		b[number][j] = pre;
		index[number]++;
	}
	int k = 0;
	for( int i=0; i<10; i++){
		for(int j = 0; j<index[i]; j++){
			a[k] = b[i][j];
			k++;	
		}
	}
}

int main(){
	cout << "Enter the numbers to be sorted: " << endl;
	int n;
	cin >> n;
	double a[n];

	for(int i=0;i<n;i++){
		cin >> a[i];
	}

	counting_sort(a, n);
	cout << "Sorted array is: " << endl;
	for(int i =0; i<n; i++){
		cout << a[i] << " ";
	}
	cout << endl;
	return 0;
}